File: haskell/nucleotide-count/DNA.hs
module DNA ( count, nucleotideCounts ) where
import qualified Data.Map.Strict as M 

nucleotideCounts :: String -> M.Map Char Int 
nucleotideCounts  st = M.unionWith ( + ) 
                       ( M.fromList [('A', 0), ('T', 0), ('C', 0), ('G', 0)] )
                       ( M.fromListWith ( + ) . flip zip ( repeat 1 ) $ st   )

count :: Char -> String -> Int
count x xs = case x `elem` [ 'A', 'C', 'T', 'U' ] of 
               True -> length . filter ( == x ) $ xs 
               _  -> error $ "invalid nucleotide " ++ show x
 

File: ocaml/nucleotide-count/dna.ml
open Core.Std

let count (xs : string) (x : char) : int =
  String.filter ~f:(fun y ->  y = x) xs
  |> String.length

let nucleotide_counts (xs : string) : int Char.Map.t =
  [('A', count xs 'A'); ('C', count xs 'C'); ('G', count xs 'G'); ('T', count xs 'T')]
  |> List.filter ~f:(fun (_, y) -> y <> 0)
  |>  Char.Map.of_alist_exn
